core/tests/unit/ldml/test_transforms.cpp

```diff
@@ -828,18 +855,65 @@ int test_normalize() {
     const std::u32string src    = U"9ce\u0300\uFFFF\u0008[\\u0001-\\uD7FE]\u0320\uFFFF\u0008\\u0001";
     const std::u32string expect = U"9ce\uFFFF\u0008[\\u0001-\\uD7FE]\u0320\u0300\uFFFF\u0008\\u0001";
     std::u32string dst = src;
-    assert(normalize_nfd_markers(dst, map, regex_sentinel));
+    assert(normalize_nfd_markers_segment(dst, map, regex_sentinel));
     if (dst != expect) {
       std::cout << "dst: " << Debug_UnicodeString(dst) << std::endl;
       std::cout << "exp: " << Debug_UnicodeString(expect) << std::endl;
     }
     zassert_string_equal(dst, expect);
     assert_equal(map.size(), 2);
-    assert_equal(map[0x0320], LDML_MARKER_ANY_INDEX);
-    assert_equal(map[MARKER_BEFORE_EOT], 0x1L);
+    assert_marker_list_equal(map[0x0320], (marker_list){LDML_MARKER_ANY_INDEX});
+    assert_marker_list_equal(map[MARKER_BEFORE_EOT], (marker_list){0x1L});
+  }
+
+  {
+    // from tests
+    marker_map map;
+    std::cout << __FILE__ << ":" << __LINE__ << "   - complex test 10 stack o' 2x2" << std::endl;
+    const std::u32string src    = U"9ce\u0300\uFFFF\u0008\u0002\uFFFF\u0008\u0002\u0320";
+    const std::u32string expect = U"9ce\uFFFF\u0008\u0002\uFFFF\u0008\u0002\u0320\u0300";
+    std::u32string dst = src;
+    assert(normalize_nfd_markers_segment(dst, map));
+    if (dst != expect) {
+      std::cout << "dst: " << Debug_UnicodeString(dst) << std::endl;
+      std::cout << "exp: " << Debug_UnicodeString(expect) << std::endl;
+    }
+    assert_equal(map.size(), 1);
+    assert_marker_list_equal(map[0x0320], ((marker_list){0x2L, 0x2L}));
+    zassert_string_equal(dst, expect);
   }
 
 
+  {
+    // from tests
+    marker_map map;
+    std::cout << __FILE__ << ":" << __LINE__ << "   - complex test 10 stack o' 2x1x2" << std::endl;
+    const std::u32string src    = U"9ce\u0300\uFFFF\u0008\u0002\uFFFF\u0008\u0001\uFFFF\u0008\u0003\u0320";
+    const std::u32string expect = U"9ce\uFFFF\u0008\u0002\uFFFF\u0008\u0001\uFFFF\u0008\u0003\u0320\u0300";
+    std::u32string dst = src;
+    assert(normalize_nfd_markers_segment(dst, map));
+    if (dst != expect) {
+      std::cout << "dst: " << Debug_UnicodeString(dst) << std::endl;
+      std::cout << "exp: " << Debug_UnicodeString(expect) << std::endl;
+    }
+    zassert_string_equal(dst, expect);
+    assert_equal(map.size(), 1);
+    assert_marker_list_equal(map[0x0320], ((marker_list){0x2L, 0x1L, 0x3L}));
+  }
+
+
+  {
+    marker_map map;
+    std::cout << __FILE__ << ":" << __LINE__ << "   - dup-char test" << std::endl;
+    const std::u32string src = U"a\uFFFF\u0008\u0001\u0300e\uFFFF\u0008\u0002\u0300";
```

Comment:
as it says, normalize_nfd_markers() would get this wrong, because U+0300 appears twice. But at least the segment code handles it properly.